Sources/AsyncHTTPClient/ConnectionPool/HTTPRequestStateMachine.swift

```diff
@@ -0,0 +1,467 @@
+            return .sendRequestHead(head, startBody: true, startReadTimeoutTimer: nil)
+        } else if head.headers.contains(name: "transfer-encoding") {
+            self.state = .running(.streaming(expectedBodyLength: nil, sentBodyBytes: 0, producer: .producing), .initialized)
+            return .sendRequestHead(head, startBody: true, startReadTimeoutTimer: nil)
+        } else {
+            self.state = .running(.endSent, .initialized)
+            return .sendRequestHead(head, startBody: false, startReadTimeoutTimer: self.idleReadTimeout)
+        }
+    }
+
+    mutating func requestVerificationFailed(_ error: Error) -> Action {
+        guard case .running(.verifyRequest, .initialized) = self.state else {
+            preconditionFailure("Invalid state")
+        }
+
+        self.state = .failed(error)
+        return .failRequest(error, closeStream: false)
+    }
+
+    mutating func requestStreamPartReceived(_ part: IOData) -> Action {
+        switch self.state {
+        case .initialized:
+            preconditionFailure("Invalid state: \(self.state)")
+
+        case .running(.verifyRequest, _),
+             .running(.endSent, _):
+            preconditionFailure("Invalid state: \(self.state)")
+
+        case .running(.streaming(let expectedBodyLength, var sentBodyBytes, let producerState), let responseState):
+            // More streamed data is accepted, even though the producer should stop. However
+            // there might be thread syncronisations situations in which the producer might not
+            // be aware that it needs to stop yet.
+
+            if let expected = expectedBodyLength {
+                if sentBodyBytes + part.readableBytes > expected {
+                    let error = HTTPClientError.bodyLengthMismatch
+
+                    switch responseState {
+                    case .initialized, .receivingBody:
+                        self.state = .failed(error)
+                    case .endReceived:
+                        #warning("TODO: This needs to be fixed. @Cory: What does this mean here?")
+                        preconditionFailure("Unimplemented")
+                    }
+
+                    return .failRequest(error, closeStream: true)
+                }
+            }
+
+            sentBodyBytes += part.readableBytes
+
+            let requestState: RequestState = .streaming(
+                expectedBodyLength: expectedBodyLength,
+                sentBodyBytes: sentBodyBytes,
+                producer: producerState
+            )
+
+            self.state = .running(requestState, responseState)
+
+            return .sendBodyPart(part)
+
+        case .failed:
+            return .wait
+
+        case .finished:
+            // a request may be finished, before we send all parts. We may still receive something
+            // here because of a thread race
+            return .wait
+        }
+    }
```

Comment:
I really think we should split this function up a bit. This is just personal opinion, but IMHO each case in a switch should have a few lines maximum, otherwise it can quickly become difficult to read and maintain.

Reply:
My feeling is that this function is only massive because of the comments now. Do you still think we should split up?

Reply:
IMHO I think this is fine as is, the amount of code in each case is pretty small.

Reply:
Yeh this is fine now